### PR TITLE
Make Cursor.Visit compatible to Go >=1.6

### PR DESCRIPTION
**Pull request by zimmski**
Starting with Go 1.6 we can no longer forward a Go pointer to C
https://golang.org/doc/go1.6#cgo There is an example fix using a lookup
table https://github.com/golang/go/wiki/cgo#function-variables which can
be used to work around this limitation.

PS: I know I have to hop back on the go-clang project, I am working on it. Please be patient. Starting October I will work full-time on a certain project which includes working on go-clang.

### CODE REVIEW

File: clang/cursor.go [hunk 1]
```diff
@@ -45,20 +46,54 @@ func (c Cursor) PlatformAvailability(availabilitySize int) (always_deprecated bo
  */
 type CursorVisitor func(cursor, parent Cursor) (status ChildVisitResult)
 
+type functionsMapping struct {
```
Comment (sbinet) [inline clang/cursor.go]: `s/functionsMapping/funcsMap/` ?
or `visitorRegistry` ?

(apologies for the bikeshedding)

File: clang/cursor.go [hunk 2]
```diff
@@ -45,20 +46,54 @@ func (c Cursor) PlatformAvailability(availabilitySize int) (always_deprecated bo
...
+	sync.Mutex
```
Comment (sbinet) [inline clang/cursor.go]: why not a `sync.RWMutex` ?
so `lookup` could use `RLock()/RUnlock`

File: clang/cursor.go [hunk 3]
```diff
@@ -45,20 +46,54 @@ func (c Cursor) PlatformAvailability(availabilitySize int) (always_deprecated bo
...
+func (fm *functionsMapping) lookup(index int) *CursorVisitor {
+	fm.Lock()
+	defer fm.Unlock()
+
+	return fm.funcs[index]
+}
+
+func (fm *functionsMapping) unregister(index int) {
+	fm.Lock()
+	defer fm.Unlock()
```
Comment (sbinet) [inline clang/cursor.go]: is the `defer` really needed here?

File: clang/cursor.go (hunk 1 above)
Comment (zimmski) [inline clang/cursor.go]: "funcsMap" is OK. I thought that we might need this more than once. Maybe I should put it in its own file?

EDIT: Maybe a combination of both for the name "funcRegistry"?

File: clang/cursor.go (hunk 2 above)
Comment (zimmski) [inline clang/cursor.go]: done

File: clang/cursor.go (hunk 3 above)
Comment (zimmski) [inline clang/cursor.go]: It is not. I am just used to typing this all the time.

done

File: clang/cursor.go (hunk 1 above)
Comment (sbinet) [inline clang/cursor.go]: let's do that when it is really needed and referenced elsewhere.

Comment (zimmski) [inline clang/cursor.go]: Alright. Is "funcRegistry" OK?

File: clang/cursor.go [hunk 4]
```diff
@@ -45,20 +46,54 @@ func (c Cursor) PlatformAvailability(availabilitySize int) (always_deprecated bo
...
+
+var visitors = &functionsMapping{
+	funcs: map[int]*CursorVisitor{},
+}
+
 // GoClangCursorVisitor calls the cursor visitor
 //export GoClangCursorVisitor
-func GoClangCursorVisitor(cursor C.CXCursor, parent C.CXCursor, cfct unsafe.Pointer) (status ChildVisitResult) {
-	fct := *(*CursorVisitor)(cfct)
+func GoClangCursorVisitor(cursor C.CXCursor, parent C.CXCursor, cfct *C.int) (status ChildVisitResult) {
```
Comment (sbinet) [inline clang/cursor.go]: isn't this a bit of pointer aliasing? (and thus not safe against "strict aliasing rules" ?)
my `C/C++` is admittedly a bit rusty, but I think we could still keep the `cfct unsafe.Pointer` and cast it to a `*C.int`:

``` go
fct := *(*C.int)(cfct)
```

File: clang/cursor.go (hunk 1 above)
Comment (sbinet) [inline clang/cursor.go]: sure.

Comment (zimmski) [reply clang/cursor.go]: done

File: clang/cursor.go (hunk 4 above)
Comment (zimmski) [reply clang/cursor.go]: You are right, it is!

done

**Comment by sbinet** [comment]
thanks.
LGTM :)

**Comment by zimmski** [comment]
I have a day off, so I thought... go-clang

**Pull request merged**